### PR TITLE
Add support for strict mode

### PR DESCRIPTION
**Pull request by shadowspawn**
Minimist does not support validating that the command-line arguments match what the author intended. The default approach is to always parse the command-line arguments, and it is difficult for the author to detect problems. This can lead to a poor end-user experience when unintended usage silently produces unexpected results.

This PR adds `opts.strict` which will throw for these usage errors:
- "string" option used without a value
- "boolean" option given a value (other than 'true' or `false')
- unknown option
- "number" option (new) used without a number value

See: #37 #39

### CODE REVIEW

File: test/array.js [hunk 1]
```diff
@@ -70,53 +70,3 @@ test('auto bool accumulates with auto string', function (t) {
 
 	t.end();
 });
-
-test('declared boolean overwrites string', function (t) {
-	var options = {
```
Comment (ljharb) [inline test/array.js]: why remove these tests?

Comment (shadowspawn) [reply test/array.js]: The (very edge) scenario the tests were checking is no longer reproducible.

I had previously written a weird and wonderful test to cover an edge case where the author declared a boolean option and the command-line arguments stored a string [sic] and then also stored a boolean. The test was checking the end result was storing a boolean rather than triggering the implicit array handling.

Previously this was technically possible through misuse of a short option like `-b=xyz` storing `b: 'xyz'`. This PR makes the short and long option processing consistent, so `-b=xyz` now stores `b:true` if `b` is boolean.

Comment (ljharb) [reply test/array.js]: Isnt that a breaking change?

Comment (shadowspawn) [reply test/array.js]: I'm currently treating it as a bonus bug fix. The behaviour was not being explicitly tested. (I have added some new tests in this PR to nail it down for the future.)

I did change the behaviour somewhat accidentally while unifying the short and long boolean behaviours to add strict mode in one place instead of many. I'm happy with how it is turning out so far, but still a work in progress.

**Comment by shadowspawn** [comment]
More detail about interaction of `opts.unknown` and strict mode. The README says:

> `opts.unknown` - a function which is invoked with a command line parameter not defined in the `opts` configuration object. If the function returns `false`, the unknown option is not added to `argv`.

<s>
Since there is specific mention of returning `false` to effectively ignore the option, I have left that particular behaviour in place. Any other return value from `opts.unknown will fall through to the normal strict processing and throw for an unknown option.

Two other approaches which I considered:
- `opt.unknown` could override the strict checking for unknown options completely. However, `opt.unknown` is currently a bit messy and difficult to use. I don't think it is suitable as an override of strict mode for unknown options. (e.g. does not get called for short options without a value, passes in argument rather than key, gets called for positional arguments. Yuck.)
- Could call `unknown` as now, but do the strict checks whatever is returned. Independent.
</s>

Due to code refactor to avoid changing non-strict behaviour, currently independent. Still calling `opts.unknown` in strict mode, but the return value does not affect the strict checks.

**Comment by ljharb** [comment]
I’d like to understand any differences from util.parseArgs here - it’d be ideal to match the api as much as possible.

**Comment by shadowspawn** [comment]
I am keeping `util:parseArgs` in mind, but there are lots of differences. Are there particular aspects you had in mind? 

Sort of the inverse of what you asked but easier, demo of current state of this PR and `parseArgs` where they are morally similar. (I didn't copy the error messages but not against that, just didn't do it.)

```js
// minimist.js
const parseArgs = require('minimist');
try {
    const result = parseArgs(process.argv.slice(2), {
        strict: true,
        string: ['str'],
        boolean: ['bbb']
    });
    console.log(result);
} catch (err) {
    console.log(err.message);
}
```

```console
$ node minimist.js --oops
Unknown option "oops"
$ node minimist.js --str 
Missing option value for option "str"
$ node minimist.js --bbb=xyz
Unexpected option value for option "bbb"
```

```js
// parseArgs.js
const { parseArgs } = require('node:util');
try {
    const result = parseArgs({
        strict: true,
        options: {
            str: { type: 'string' },
            bbb: { type: 'boolean'}
        }
    });
    console.log(result);
} catch (err) {
    console.log(err.message);
}
```

```console
$ node parseArgs.js --oops   
Unknown option '--oops'
$ node parseArgs.js --str 
Option '--str <value>' argument missing
$ node parseArgs.js --bbb=xyz
Option '--bbb' does not take an argument
```

**Comment by shadowspawn** [comment]
I currently have added a `known` property in this PR,  but now thinking instead of adding a `number` property. I'll try it out before justifying it...

File: test/array.js (hunk 1 above)
Comment (shadowspawn) [reply test/array.js]: The change isn't essential to the implementation, and could preserve legacy inconsistent behaviour for short flags.

Comment (shadowspawn) [reply test/array.js]: I have refactored to preserve the previous (non-strict) behaviour by preserving the original code flow! Keep it simple stupid.

**Review by Karimsultan — APPROVED** [review]

**Comment by shadowspawn** [comment]
This didn't work out as well as I hoped. I have some ideas about a lighter weight approach to strict which I might try in future. More a "stricter" than a "strict", since minimist has lots of automatic behaviours and options to tweak the behaviours, but is neither strict in how it parses nor what types it produces in results.
